expan/core/statistical_test.py

```diff
@@ -19,6 +19,10 @@ class StatisticalTest(JsonSerializable):
     :type  variants: Variants
     """
     def __init__(self, data, kpi, features, variants):
+        # sometimes we want to postpone the actual creation of the
+        # dataframe
+        if callable(data):
+            data = data()
```

Comment:
I am still not sure if you get my point. 

Given a feature with multiples values, say "Browser" which have "Chrome", "FF" and "IE". This will create 3 `StatisticalTest` objects where you pass the same `callable` function to it. 
"StatisticalTest(bla, browser=chrome)"
"StatisticalTest(bla, browser=ff)"
"StatisticalTest(bla, browser=ie)"

From our experience, we know this `callable` function to create dataframe is runtime-expensive (groupby, multi-variance-drop, outlier-filtering, etc). We want to call it only once for the feature "browser" (instead of repeating 3 times for each browser value). 

how can you ensure that?
is it the smart counter example we discussed? maybe I didn't understand your point.

Currently, we created dfs once per subgroup. And pass the reference when creating

Reply:
The current implementation in `ExpanService` computes the dataframe only once per subgroup. One for `browser`, one for `date`, one for `appdomain` and so on.

Reply:
@shansfolder thanks for the detailed rebuttal.

This is only the Expan side of the problem. On the client side (ExpanService), we could make `data`, for example, caching:

```
def dataframe_maker():
  df = [None] # we need a list here because of how closures in Python work
  def make_df():
    if not df[0]:
      df[0] = load_dataframe_do_group_by_etc(args)
    return df[0]
  return make_df
```

And then pass it as `df_maker = dataframe_maker(args)` and then use it in different `StatisticalTest`s as

```
# in one place
df1 = df_maker()

# in some other place
df2 = df_maker()

# in a third place
df3 = df_maker()
```

The logic above would ensure that `load_dataframe_do_group_by_etc(args)` is called only once despite `df_maker()` being called three times.

What do you think of it?